app/processing/siti-agri/handle-known-defects/cap-bps-penalties/recalculate-bps-penalties.js

```diff
@@ -0,0 +1,34 @@
+const recalculateBPSPenalties = (paymentRequest) => {
+  const schemeCodes = [...new Set(paymentRequest.invoiceLines.map(invoiceLine => invoiceLine.schemeCode))]
+
+  schemeCodes.forEach((schemeCode) => {
+    const invoiceLinesByScheme = paymentRequest.invoiceLines.filter(invoiceLine => invoiceLine.schemeCode === schemeCode)
+    let grossAfterPenalties = calculateGrossAfterPenalties(invoiceLinesByScheme)
+
+    if (grossAfterPenalties < 0) {
+      const p04Penalty = invoiceLinesByScheme.filter(invoiceLine => invoiceLine.description.match(/^P04/gm))[0]
+      const p02Penalty = invoiceLinesByScheme.filter(invoiceLine => invoiceLine.description.match(/^P02/gm))[0]
+
+      if (p04Penalty) {
+        p04Penalty.value = p04Penalty.value - grossAfterPenalties > 0 ? 0 : p04Penalty.value -= grossAfterPenalties
+        grossAfterPenalties = calculateGrossAfterPenalties(invoiceLinesByScheme)
+      }
+
+      if (grossAfterPenalties < 0) {
+        p02Penalty.value -= grossAfterPenalties
```

Comment:
If there is no P02 this would throw an error

Reply:
If that happens doesnt it mean something has gone wrong though at a wider level, because then the gross would still be less than 0 but there would be no other penalties to reduce? So what should happen in this scenario. I was thinking of just amending to 
`if (grossAfterPenalties < 0 && p02Penalty)` 
but that wouldnt really solve the problem as the payment request would then be returned and there would be a discrepency between header value and line total so it would be rejected?
Should we return a specific error? or wrap it in a try/catch?

Reply:
That's true but it would be caught by the validation steps later and alert would be sent to core solutions team.  if we throw an error, then to process just fails.  Not sure if an alert is sent in that instance?

Reply:
agree, have updated.